### PR TITLE
Implement NonZero shape inference

### PR DESCRIPTION
**Pull request by impactaky**

### CODE REVIEW

**Review by gramalingam — APPROVED** [review]

**Comment by gramalingam** [comment]
This seems similar to https://github.com/onnx/onnx/pull/3207 ... but has extra test-cases, so makes sense to pick this one. The other one also has some signoff issues preventing a merge.

**Comment by impactaky** [comment]
Thank you!

**Review by askhade — APPROVED** [review]

**Comment by askhade** [comment]
running some validations as part of the release validation work... will check this in a few

File: onnx/defs/tensor/old.cc [hunk 1]
```diff
@@ -1940,6 +1940,14 @@ ONNX_OPERATOR_SET_SCHEMA(
             "Constrain to all tensor types.")
         .TypeAndShapeInferenceFunction([](InferenceContext& ctx) {
           updateOutputElemType(ctx, 0, TensorProto::INT64);
+          TensorShapeProto output_shape;
+          auto* dim = output_shape.add_dim();
+          if (hasInputShape(ctx, 0)) {
+            const TensorShapeProto& input_shape = getInputShape(ctx, 0);
+            dim->set_dim_value(input_shape.dim_size());
+          }
+          output_shape.add_dim();
```
Comment (mkbhanda) [inline onnx/defs/tensor/old.cc:1949]: ditto, a second call to add_dim()

**Comment by mkbhanda** [comment]
What is the policy in project ONNX on branches out-of-da†e, this is of†en in busy projects. https://github.community/t/requiring-up-to-date-branches-in-a-busy-repo/1550/2

File: onnx/defs/tensor/defs.cc [hunk 2]
```diff
@@ -2694,6 +2694,14 @@ ONNX_OPERATOR_SET_SCHEMA(
             "Constrain to all tensor types.")
         .TypeAndShapeInferenceFunction([](InferenceContext& ctx) {
           updateOutputElemType(ctx, 0, TensorProto::INT64);
+          TensorShapeProto output_shape;
+          auto* dim = output_shape.add_dim();
+          if (hasInputShape(ctx, 0)) {
+            const TensorShapeProto& input_shape = getInputShape(ctx, 0);
+            dim->set_dim_value(input_shape.dim_size());
+          }
+          output_shape.add_dim();
```
[the remark this answers is not in the document]
Comment (askhade) [reply onnx/defs/tensor/defs.cc:2703]: The shape of non zero op is [rank of input, num of non zero indices] hence the 2 calls to add_dim

File: onnx/defs/tensor/old.cc (hunk 1 above)
Comment (mkbhanda) [reply onnx/defs/tensor/old.cc:1949]: Thank you for explaining @impactaky and @askhade ! Perhaps add the above explanation as a comment :-)

**Pull request merged**